RI: Discover - fix follow action

ashiagr opened the pull request:
Parent Issue: #12028

### Description

~~Follow Site (from more menu) logic in `ReaderPostListFragment` is extracted to `ReaderPostFollowUseCase` and similar to `ReaderPostBookmarkUseCase` in [Issue/12028 save for later #12400 ](https://github.com/wordpress-mobile/WordPress-Android/pull/12400), different `LiveData` instances are used to communicate data like snackbar message or refreshing post on failure in the legacy post adapter. `ReaderPostFollowUseCase` is used on discover tab VM as well as `ReaderPostListFragment`'s VM.~~

### Suggestions
1. We might want to use single `Livedata` (of generic type) in the usecase, observing it in the VM and updating respective `LiveData` instances in the VM based on result type.

2. In the future, we might also want to consider using [Flow](https://codelabs.developers.google.com/codelabs/advanced-kotlin-coroutines/#13) in the usecases when it is released as stable version.

    > Using Flow, it's natural to collect data in the ViewModel, Repository, or other data layers when needed.

    > Since Flow is not tied to the UI, you don't need a UI observer to collect a flow. This is a big difference from LiveData which always requires a UI-observer to run. It is not a good idea to try to observe a LiveData in your ViewModel because it doesn't have an appropriate observation lifecycle.

UPDATE
- As per last discussion, used `Flow` in the useCase - b331349
- Flow values are collected and appropriate livedata are updated in the handler - e74bb0a
- Also utilised notifications useCase for enabling notification on follow. - ab3956d
- Utilised isExternalFeed method to display 
    - notifications menu in the more popup - 80c4fd6
    - enable notifications snackbar - 8a48dd7

### To test

**Following -> Follow**
1. Open Reader - Following tab
2. Click "Following" in the more menu on a followed site post
3. Notice  in the more menu: "Following" (in green) changes to "Follow" (in blue) 

**Follow -> Following + Snackbar to enable Notifications**
1. Continue from above
2. Click "Follow" in the more menu
3. Notice Snackbar is shown
4. Click on "Enable" action in the snackbar
5. Notice  in the more menu 
    5a - Follow (in blue) changes to Following (in green)
    5b - "Turn off site notifications" is shown (in green)

Repeat above steps for an external feed like https://timesofindia.indiatimes.com/rssfeedstopstories.cms. Notice that Enable notifications snackbar is not shown.

**Network Failure**
1. Continue from above
2. Enable Airplane mode
3. Click "Following" in the more menu
4. Notice that a snackbar is shown with an error message for no network

**Request Failure**
1. Make changes in the code [here](https://github.com/wordpress-mobile/WordPress-Android/pull/12667/files#diff-9758e44dc0bc3288df32504d25282546R48) to return `RequestFailed` state
2. Reinstall app -> Open Reader -> Following tab
3. Click "Following" in the more menu on a followed site post
4. Notice in the more menu: "Following" (in green) changes to "Follow" (in blue) 
5. Wait for few secs
6. Notice in the more menu: 
    6a - Follow (in blue) changes to Following (in green)
    6b - Snackbar is shown with an error message

Above tests should also work for discover tab (with RI FF enabled).

### Notes
~~1. Targeting feature branch `issue/12028-reader-post-action-usecases` instead of `develop`  till we reach an agreement on final design for usecases.~~
~~2. Above tests might not work as expected if the site to be followed has a wrong `blogId` set in the `tbl_blog_info` table. This is due to an existing bug where `blogId` is wrongly set to `feedId` for a blog followed as a feed. `ReaderBlog` doesn't parse `blog_ID` key [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/models/ReaderBlog.java#L66-L67). This issue is being fixed separately.~~. Fixed in #12694
3. Have not entirely removed [these methods](https://github.com/wordpress-mobile/WordPress-Android/blob/21d5e25380bdbf67f7e65ee6a38950ce44c322e2/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java#L2619-L2656) extracted to `ReaderPostFollowUseCase` as they're also used on Site Details -> `ReaderSiteHeaderView` and `ReaderPostDetailFragment` -> `ReaderPostDetailHeaderView`, `ReaderSimplePostView`.
~~4. Extracted logic needs further refactoring.~~
5. Will add tests once these changes are approved (Added "Not Ready for Merge" label till then). 

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Update SnackbarMessageHolder to accept a string

ashiagr committed:
Add follow post use case

ashiagr committed:
Observe data from follow use case

ashiagr committed:
Fix naming

ashiagr committed:
Subscribe to OnSubscriptionUpdated

ashiagr committed:
Merge branch 'develop' into issue/12028-fix-follow-action

ashiagr committed:
Refresh post status on legacy post adapter

ashiagr committed:
Rename method

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/pages/SnackbarMessageHolder.kt:
```diff
@@ -2,9 +2,12 @@ package org.wordpress.android.ui.pages
 
 import androidx.annotation.StringRes
 
+const val INVALID_MESSAGE_RES = -1
```
I'm wondering if we could use UIString instead of using an invalid id, wdyt?

ashiagr replied:
Converting to UIString looks like a great idea! I'll create a separate PR as it will impact several classes.
I'll convert both messageRes and buttonTitleRes to UIString.

ashiagr replied:
Converted here: #12740

ashiagr committed:
Merge branch 'develop' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverFragment.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostListViewModel.kt

ashiagr committed:
Fix merge

ashiagr committed:
Merge branch 'issue/12028-notifications-usecase' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostListViewModel.kt

ashiagr committed:
Merge branch 'issue/12028-notifications-usecase' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt

ashiagr committed:
Update usecase to only include logic to follow a site

Also
- Removed logic to enable notifications. It will be included in the action handler via notifications usecase.
- Used flow to emit results
- Used wrappers for table and action

ashiagr committed:
Collect values from follow use case and update LiveData instances

ashiagr committed:
Show enable notifications snackbar on site follow

If the user clicks enable, notifications are enabled using the notifications usecase

ashiagr committed:
Use utility method isExternalFeed to decide site notification menu display

ashiagr committed:
Rename follow usecase

ashiagr committed:
Remove unused method

ashiagr committed:
Add external feed check for enable notifications snackbar

ashiagr committed:
Merge branch 'develop' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt

ashiagr committed:
Optimise imports

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -477,6 +478,8 @@ public void onCreate(Bundle savedInstanceState) {
                 })
         );
 
+        mViewModel.getRefreshPost().observe(getViewLifecycleOwner(), this::setFollowStatusForBlog);
```
Hmm, I'm wondering if this isn't a bit misleading. We are observing "refreshPost" event but we are calling "setFollowStatusForBlog" -> we technically shouldn't know that the VM wanted us to update the follow status. Moreover, if I as a future developer see `refreshPost` livedata in the VM and I use it I expect that the posts will get refreshed. However, it won't be the case. 
Would it work if we invoked `getPostAdapter().refresh()` instead of `setFollowStatusForBlog(..)`? If not, I'd consider renaming this LiveData to "updateFollowStatus" (the livedata in the handler could be called "followStatusUpdated"), wdyt?

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderSiteFollowUseCase.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    sealed class FollowSiteState {
+        data class ReaderPostData(
```
We might want to consider renaming this to something like "PostFollowStatusChanged". Since "ReaderPostData" technically isn't a state, yet it extends `FollowSiteState`. Wdyt?

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt:
```diff
@@ -194,6 +235,35 @@ class ReaderPostCardActionsHandler @Inject constructor(
     }
 
+    private fun prepareEnableNotificationSnackbarAction(blogName: String?, blogId: Long): () -> Unit {
+        return {
```
Feel free to ignore this: This will be a bit difficult to test. I'd consider moving the creation of `notificationMessage` into a different class. Wdyt?

ashiagr replied:
Extracted to different class here d3795d8.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt:
```diff
@@ -194,6 +235,35 @@ class ReaderPostCardActionsHandler @Inject constructor(
+                                        analyticsTrackerWrapper
+                                                .track(FOLLOWED_BLOG_NOTIFICATIONS_READER_ENABLED, blogId)
+                                        siteNotificationsUseCase.updateSubscription(blogId, NEW)
+                                        siteNotificationsUseCase.updateBlogInDb(blogId, false)
```
Since we are accessing the db here, I'd consider launching a coroutine on a bg thread here. Wdyt?

ashiagr replied:
Agree, done in b66b46d

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostMoreButtonUiStateBuilder.kt:
```diff
@@ -39,8 +41,8 @@ class ReaderPostMoreButtonUiStateBuilder @Inject constructor(
-            // When blogId and feedId are not equal, post is not a feed so show notifications option.
-            if (post.blogId != post.feedId) {
+            // When post not from external feed so show notifications option.
+            if (!readerUtilsWrapper.isExternalFeed(post.blogId, post.feedId)) {
```
Nice catch :)

malinajirka reviewed (commented):
Thanks @ashiagr! I've tested the changes and found one super minor issue - we can either ignore it or create a low priority ticket for groundskeeping (develop behaves the same way).

1. Follow a post
2. Enable notificaitons
3. Unfollow the post
4. Follow the post
5. Notice the snackbar with "enable" is shown even though the notifs are already enabled
- the question is whether we shouldn't show the snackbar in this scenario or if we should disable notifications when you unfollow a site.

I have left some in-code comments. None of them is important - let me know what you think.

ashiagr committed:
Rename variables

ReaderPostCardActionsHandler.refreshPost -> ReaderPostCardActionsHandler.followStatusUpdated
ReaderPostListViewModel.refreshPost -> ReaderPostListViewModel.updateFollowStatus
FollowSiteState.ReaderPostData -> FollowSiteState.PostFollowStatusChanged

ashiagr committed:
Utilise CharSequence's isEmpty() directly

ashiagr committed:
Rename method + param to update notification enable status

ashiagr committed:
Delete site notifications subscription on site un-follow

ashiagr committed:
Launch enable notification action in default scope

ashiagr committed:
Extract notification message to utils

ashiagr committed:
Merge branch 'develop' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java
#	WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderBlogActionsWrapper.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostListViewModel.kt

ashiagr committed:
Fix merge

ashiagr commented:
Thanks for the review @malinajirka !

Regarding 

> I've tested the changes and found one super minor issue - we can either ignore it or create a low priority ticket for groundskeeping (develop behaves the same way).

Decided to fix it in this PR itself - afe628e

> the question is whether we shouldn't show the snackbar in this scenario or if we should disable notifications when you unfollow a site.

I think it would make more sense to disable site notifications once we unfollow it.

Tests will be added in a separate PR. Added it to backlog here: #12821.

Ready for another round 🙇‍♀️!

ashiagr committed:
Merge branch 'develop' into issue/12028-fix-follow-action

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostMoreButtonUiStateBuilder.kt
#	WordPress/src/test/java/org/wordpress/android/ui/reader/discover/ReaderPostMoreButtonUiStateBuilderTest.kt

ashiagr committed:
Fix merge

ashiagr committed:
Assign correct param

malinajirka commented:
Thanks for the changes @ashiagr! It seems to work as expected (from the user point of view). However, I can see an error in LogCat
1. Click on More -> follow blog
2. Click on More -> enable notifs
3. Click on More -> unfollow blog
4. See error in logcat (it doesn't occur every time, but most of the times)
<img src="https://user-images.githubusercontent.com/2261188/91832714-0856e600-ec46-11ea-8c4e-8b2f9c770269.gif" width="800px" />



```
2020-09-01 11:21:31.189 7441-7562/org.wordpress.android.prealpha E/SQLiteDatabase: Error inserting EMAIL_POSTS_FREQUENCY=null SHOULD_EMAIL_POSTS=false FEED_ID=108454787 SUBSCRIPTION_ID=609519261 BLOG_NAME=Site Title SHOULD_NOTIFY_POSTS=false BLOG_ID=182078938 URL=http://gfsdgw343ye4atgaeye5yh.car.blog SHOULD_EMAIL_COMMENTS=false
    android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SubscriptionModel.FEED_ID, SubscriptionModel.URL (code 2067)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:783)
        at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:788)
        at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:86)
        at android.database.sqlite.SQLiteDatabase.insertWithOnConflict(SQLiteDatabase.java:1548)
        at android.database.sqlite.SQLiteDatabase.insert(SQLiteDatabase.java:1417)
        at com.yarolegovich.wellsql.InsertQuery.execute(InsertQuery.java:61)
        at org.wordpress.android.fluxc.persistence.AccountSqlUtils.updateSubscriptions(AccountSqlUtils.java:147)
        at org.wordpress.android.fluxc.store.AccountStore.updateSubscriptions(AccountStore.java:1307)
        at org.wordpress.android.fluxc.store.AccountStore.onAccountAction(AccountStore.java:912)
        at org.wordpress.android.fluxc.store.AccountStore.onAction(AccountStore.java:818)
        at java.lang.reflect.Method.invoke(Native Method)
        at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:507)
        at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:501)
        at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
```

I'm not 100% sure it's related to this PR, but seems like it. Wdyt?

ashiagr commented:
>  I can see an error in LogCat
```
2020-09-01 11:21:31.189 7441-7562/org.wordpress.android.prealpha E/SQLiteDatabase: Error inserting EMAIL_POSTS_FREQUENCY=null SHOULD_EMAIL_POSTS=false FEED_ID=108454787 SUBSCRIPTION_ID=609519261 BLOG_NAME=Site Title SHOULD_NOTIFY_POSTS=false BLOG_ID=182078938 URL=http://gfsdgw343ye4atgaeye5yh.car.blog SHOULD_EMAIL_COMMENTS=false
    android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SubscriptionModel.FEED_ID, SubscriptionModel.URL (code 2067)
  
```

I tested it occurs on develop as well (with RI FF disabled). Should we create a separate issue and address it there?

malinajirka reviewed (approved):
👍 Yes, create an issue and assign it to Groundskeeping triage please. Thanks 🙇

designsimply commented:
This is awesome. Thank you for working on this!
